sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java

```diff
@@ -110,24 +117,46 @@ public EvaluationOutcome evaluate(MesosResourcePool mesosResourcePool, PodInfoBu
                         .build();
             }
 
+            Set<Protos.Volume> existingVolumes = podInfoBuilder.getTaskBuilder(taskName)
+                    .getContainerBuilder()
+                    .getVolumesList()
+                    .stream()
+                    .collect(Collectors.toSet());
+            logger.debug("Existing volumes for {}: {}",
+                        taskName,
+                        existingVolumes.stream().map(v -> v.getContainerPath()).toArray());
+
+            Set<Protos.Volume> additionalVolumes = getExecutorInfoSecretVolumes(
+                    transportEncryptionSpec,
+                    tlsArtifactPaths
+            );
+            logger.debug("Required volumes for {}: {}",
+                        taskName,
+                        additionalVolumes.stream().map(v -> v.getContainerPath()).toArray());
+
+            if (additionalVolumes.removeAll(existingVolumes)) {
+                logger.debug("Duplicate volumes for {} removed. Remaining: {}",
+                            taskName,
+                            additionalVolumes.stream().map(v -> v.getContainerPath()).toArray());
+            }
+
             // Share keys to the task container
             podInfoBuilder
                     .getTaskBuilder(taskName)
                     .getContainerBuilder()
-                    .addAllVolumes(getExecutorInfoSecretVolumes(transportEncryptionSpec, tlsArtifactPaths));
-
+                    .addAllVolumes(additionalVolumes);
         }
 
         return EvaluationOutcome.pass(this, "TLS certificate created and added to the task").build();
     }
 
-    private static Collection<Protos.Volume> getExecutorInfoSecretVolumes(
+    private static Set<Protos.Volume> getExecutorInfoSecretVolumes(
             TransportEncryptionSpec spec, TLSArtifactPaths tlsArtifactPaths) {
-        Collection<Protos.Volume> volumes = new ArrayList<>();
-        for (TLSArtifactPaths.Entry entry : tlsArtifactPaths.getPathsForType(spec.getType(), spec.getName())) {
-            volumes.add(getSecretVolume(entry));
-        }
-        return volumes;
+
+        Collection<TLSArtifactPaths.Entry> paths = tlsArtifactPaths.getPathsForType(spec.getType(), spec.getName());
+        return paths.stream()
+                .map(TLSEvaluationStage::getSecretVolume)
+                .collect(Collectors.toSet());
```

Comment:
Does this set need a custom Comparator implementation? Example of building a set with a custom comparator from a stream: https://github.com/mesosphere/dcos-commons/pull/2571/files#diff-d71effdf03d955bff6c0fe818ae25c0cR227

Reply:
It seems as if the `Protos.Volume` objects being compared here behave well enough. Could be that the protobuf compiler already generates applicable comparators for equality.

If your asking whether we should *only* be comparing the volume paths, that's a fair point, but if we're trying to map different secrets to the same path, then that's also an issue.

Reply:
Oh, that makes sense. For some reason I forgot that the protobuf types included equality checks.